### DISCUSSION TITLE
How to improve performance with complex aqueous solutions?

### DISCUSSION THREAD

**Discussion by martinvoigt** [reaktoro/reaktoro#287] [General]
I noticed that the performance of the equilibrium solver is quite slow when dealing with solutions that contain a bunch of elements.
See an example below (related to a [previous question](https://github.com/reaktoro/reaktoro/discussions/285#discussion-4913701)), which compares it with PHREEQC (using the IPhreeqcPy module) as example.

On my computer, the Reaktoro code takes about a minute, while PHREEQC takes less than 10 seconds.
How do I properly take advantage of Reaktoro to do these calculations faster than PHREEQC?

Test data:
```python
import numpy as np
import pandas as pd
from IPhreeqcPy import IPhreeqc
import reaktoro as rkt

testdata = pd.DataFrame()
testdata['temp'] = 10 + np.random.rand(1000)
testdata['co2'] = 0.02 * np.random.rand(1000)
```

Reaktoro test:
```python
db = rkt.PhreeqcDatabase('llnl.dat')

solution = rkt.AqueousPhase(rkt.speciate("H O C Ca Fe K Mg Na Si S Cl F"))
solution.setActivityModel(rkt.chain(
    rkt.ActivityModelDebyeHuckelPHREEQC(),
    rkt.ActivityModelDrummond("CO2")
))

gases = rkt.GaseousPhase("CO2(g) H2O(g)")
gases.setActivityModel(rkt.ActivityModelRedlichKwong())

system = rkt.ChemicalSystem(db, solution, gases)

waterState = rkt.ChemicalState(system)
waterState.set("H2O", 1.0, "kg")
waterState.set("Ca+2", 10, "mg")
waterState.set("Fe+2", 0.01, "mg")
waterState.set("K+", 1, "mg")
waterState.set("Mg+2", 5, "mg")
waterState.set("Na+", 10, "mg")
waterState.set("SiO2", 50, "mg")
waterState.set("SO4-2", 10, "mg")
waterState.set("CO2", 50, "mg")
waterState.set("Cl-", 100, "mg")
waterState.set("F-", 2, "mg")

specs = rkt.EquilibriumSpecs(system)
specs.temperature()
specs.pressure()
specs.pH()

solver = rkt.EquilibriumSolver(specs)

conditions = rkt.EquilibriumConditions(specs)
conditions.temperature(7.2, "celsius")
conditions.pressure(1.0, "atm")
conditions.pH(8.0)

solver.solve(waterState, conditions)

specs2 = rkt.EquilibriumSpecs(system)
specs2.temperature()
specs2.phaseAmount("GaseousPhase")

solver2 = rkt.EquilibriumSolver(specs2)

conditions2 = rkt.EquilibriumConditions(specs2)
conditions2.phaseAmount("GaseousPhase", 1, "umol")
conditions2.setLowerBoundPressure(0.0005, "bar")
conditions2.setUpperBoundPressure(200.0, "bar")

def ReaktoroTest(r):
    conditions2.temperature(r['temp'], "celsius")

    state2 = rkt.ChemicalState(waterState)
    state2.add("CO2", r['co2'], "kg")
    solver2.solve(state2, conditions2)

    P = float(state2.pressure()) * 1e-5

    return P

start_time = pd.Timestamp.now()
testdata.apply(ReaktoroTest, axis=1)
print(pd.Timestamp.now()-start_time)
```

PHREEQC example:
```python
p = IPhreeqc()
p.LoadDatabase('llnl.dat')

def PHREEQCTest(r):
    p.RunString(
    """
    CALCULATE_VALUES
    P
    -start
        10 p = PRESSURE
        20 SAVE p
    -end

    SOLUTION 1
      temp 7.2
      pH 8
      Ca 10
      Fe 0.01
      K 1
      Mg 5
      Na 10
      Si 50 as SiO2
      S(6) 10 as SO4
      C(4) 50 as CO2
      Cl 100
      F 2

    REACTION 1
        CO2 """ + str(r['co2'] / 44 * 1e3) + """

    REACTION_TEMPERATURE
        """ + str(r['temp']) + """

    GAS_PHASE
        -fixed_volume
        -volume 1e-8
        CO2(g) 0
        H2O(g) 0

    SELECTED_OUTPUT
      -reset false
      -calculate_values P
      -high_precision true
    """
    )

    P = p.GetSelectedOutputArray()[-1][0] / 1.01325
    
    return P

start_time = pd.Timestamp.now()
testdata.apply(PHREEQCTest, axis=1)
print(pd.Timestamp.now()-start_time)
```

**Comment by allanleal** [comment]
Hi Martin, thanks for reporting this and providing a complete working example. I'll investigate what is going on. Do you know if the `IPhreeqc` object maintains state of the previous calculation and uses it as warm start for the new calculation?

**Reply by allanleal** [reply]
There is no warm starting - I checked here. I'll investigate this in the next days. For robustness reasons, Reaktoro is at the moment configured, by default, to solve the system of linear equations that arise when solving the non-linear equilibrium equations using a strategy that needs to solve matrix equations of size (N-E) by (N-E) where N is number of species and E is number of elements. I'll check an alternative strategy that reduces the matrix sizes to E x E. For larger systems, N >> E, this can make a significant difference.

**Comment by allanleal** [comment]
I profiled your example (using callgrind). The performance issue you are experiencing is unrelated to any linear algebra calculation (you can disregard my previous comment about matrix equations). Reaktoro v2 still has a temporary (and not as efficient as it could be) strategy for computing the Jacobian matrix of the system of equations that arise when solving the equilibrium problem. The activity models/equations of state assigned to the phases are called **once per species at each Newton iteration**. These repeated calls are needed to compute derivatives such as the derivative of activity of CO2 with respect to the amount of Na+. 

This is a temporary issue. A fix is possible by not computing derivatives with respect to species that play no significant role in the calculation. This would avoid many repeated calls of these activity models and equations of state. 

Also important to notice is that for equilibrium calculations in which T and P are given (in your case, P is unknown), some improvements have been made in the past. In such cases, the performance hit is not as significant (but we need an ultimate general fix for this that works for all chemical equilibrium specifications).

**Reply by martinvoigt** [reply]
Thanks for getting to the bottom of this, Allan! Looking forward to these updates :)
In the meantime, is there anything else to optimize in my script, or is this more or less the way I'm supposed to use Reaktoro for a problem like this?

**Reply by allanleal** [reply]
Hi Martin, you're doing it right. Depending on how your sequence of calculations is formulated, you may want to use the previously calculated chemical state as the initial guess for the subsequent calculation. Say, you just calculated for 10°C, and you want to repeat the calculation at 12 °C (this "warm start" strategy will produce way fewer iterations than the "cold start"). However, sometimes we want to do a sequence of calculations, say, from 10°C to 100°C, at 1 bar. And then, perform a new sequence from 10°C to 100°C, at 50 bar. In this "transition" from one sequence to another, using the final equilibrium state at 100°C and 1 bar as the initial guess for the new calculation at 10°C and 50 bar is possibly not a good choice, and it could lead to numerical issues. In this case, it is best to start with cold-start for every new sequence of calculations. I hope this clarifies!

**Reply by martinvoigt** [reply]
Thanks - sounds good!
